// @adobe/uix-host-react/dist/index.js
'use strict';

var uixHost = require('@adobe/uix-host');
var React = require('react');

function _interopDefault (e) { return e && e.__esModule ? e : { default: e }; }

var React__default = /*#__PURE__*/_interopDefault(React);

// src/index.ts
var ExtensionContext = React.createContext(
  {}
);

// src/components/Extensible.tsx
function areExtensionsDifferent(set1, set2) {
  const ids1 = Object.keys(set1).sort();
  const ids2 = Object.keys(set2).sort();
  return ids1.length !== ids2.length || ids1.some((id, index) => id !== ids2[index]);
}
function Extensible({
  appName,
  children,
  extensionsProvider,
  guestOptions,
  runtimeContainer,
  debug,
  sharedContext
}) {
  const hostName = appName || window.location.host || "mainframe";
  const [extensions, setExtensions] = React.useState({});
  const [extensionListFetched, setExtensionListFetched] = React.useState(false);
  React.useEffect(() => {
    extensionsProvider().then((loaded) => {
      setExtensions(
        (prev) => areExtensionsDifferent(prev, loaded) ? loaded : prev
      );
    }).catch((e) => {
      console.error("Fetching list of extensions failed!", e);
    }).finally(() => {
      setExtensionListFetched(true);
    });
  }, [extensionsProvider]);
  const [host, setHost] = React.useState();
  React.useEffect(() => {
    function logError(msg) {
      return (e) => {
        const error = e instanceof Error ? e : new Error(String(e));
        console.error(msg, error, extensions, guestOptions);
      };
    }
    const host2 = new uixHost.Host({ debug, hostName, runtimeContainer, sharedContext });
    setHost(host2);
    if (!Object.entries(extensions).length) {
      return;
    }
    host2.load(extensions, guestOptions).catch(logError("Load of extensions failed!"));
  }, [debug, hostName, runtimeContainer, extensions]);
  if (!host) {
    return null;
  }
  return /* @__PURE__ */ React__default.default.createElement(
    ExtensionContext.Provider,
    {
      value: {
        host,
        extensionListFetched
      }
    },
    children
  );
}
var OutsideOfExtensionContextError = class extends Error {
  constructor(msg) {
    super(msg);
    this.outsideOfExtensionContext = true;
    Object.setPrototypeOf(this, OutsideOfExtensionContextError.prototype);
  }
};
function useHost() {
  const extensionsInfo = React.useContext(ExtensionContext);
  if (!(extensionsInfo.host instanceof uixHost.Host)) {
    const error = new OutsideOfExtensionContextError(
      "Attempt to use extensions outside of ExtensionContext. Wrap extensible part of application with Extensible component."
    );
    return {
      host: void 0,
      error
    };
  }
  return { error: void 0, host: extensionsInfo.host };
}
var defaultIFrameProps = {
  width: "100%",
  height: "100%",
  style: {
    border: "none"
  }
};
var defaultSandbox = "allow-scripts allow-forms allow-same-origin";
var GuestUIFrame = ({
  guestId,
  src = "",
  onConnect,
  onDisconnect,
  onConnectionError,
  onResize,
  methods,
  sandbox = "",
  style,
  ...customIFrameProps
}) => {
  const ref = React.useRef();
  const { host } = useHost();
  if (!host) {
    return null;
  }
  const guest = host.guests.get(guestId);
  const frameUrl = new URL(src, guest.url.href);
  React.useEffect(() => {
    if (ref.current) {
      let mounted = true;
      let connection;
      const connectionFrame = ref.current;
      if (methods) {
        guest.provide(methods);
      }
      const connecting = guest.attachUI(connectionFrame);
      connecting.then((c) => {
        connection = c;
        if (!mounted) {
          c.tunnel.destroy();
        } else if (onConnect) {
          onConnect();
        }
      }).catch((error) => {
        if (mounted && !connection && connectionFrame === ref.current) {
          const frameError = new Error(
            `GuestUIFrame connection failed: ${error && error.stack || error}`
          );
          Object.assign(frameError, {
            original: error,
            ref,
            guest,
            host
          });
          if (onConnectionError)
            onConnectionError(frameError);
        }
      });
      return () => {
        mounted = false;
        if (connection) {
          connection.tunnel.destroy();
        }
      };
    }
  }, []);
  React.useEffect(() => {
    if (ref.current && onResize) {
      const currentFrame = ref.current;
      return guest.addEventListener(
        "guestresize",
        ({ detail: { guestPort, iframe, dimensions } }) => {
          if (guestPort.id === guest.id && iframe === currentFrame) {
            onResize(dimensions);
          }
        }
      );
    }
  }, [ref.current, guest.id, onResize]);
  return /* @__PURE__ */ React__default.default.createElement(
    "iframe",
    {
      ...defaultIFrameProps,
      ref,
      src: frameUrl.href,
      name: `uix-guest-${guest.id}`,
      sandbox: sandbox ? uixHost.makeSandboxAttrs(defaultSandbox, sandbox).join(" ") : defaultSandbox,
      style: style ? { ...style, ...defaultIFrameProps.style } : defaultIFrameProps.style,
      ...customIFrameProps,
      ...uixHost.requiredIframeProps
    }
  );
};
var NO_EXTENSIONS = [];
function useExtensions(configFactory, deps = []) {
  const { host, error } = useHost();
  if (error) {
    return {
      extensions: NO_EXTENSIONS,
      loading: false,
      error
    };
  }
  const baseDeps = [host, ...deps];
  const {
    requires,
    provides,
    updateOn = "each"
  } = React.useMemo(() => configFactory(host), baseDeps);
  const getExtensions = React.useCallback(() => {
    const newExtensions = [];
    const guests = host.getLoadedGuests(requires);
    for (const guest of guests) {
      if (provides) {
        guest.provide(provides);
      }
      newExtensions.push(guest);
    }
    return newExtensions.length === 0 ? NO_EXTENSIONS : newExtensions;
  }, [...baseDeps, requires]);
  const subscribe = React.useCallback(
    (handler) => {
      const eventName = updateOn === "all" ? "loadallguests" : "guestload";
      host.addEventListener(eventName, handler);
      return () => host.removeEventListener(eventName, handler);
    },
    [...baseDeps, updateOn]
  );
  const [extensions, setExtensions] = React.useState(() => getExtensions());
  React.useEffect(() => {
    return subscribe(() => {
      setExtensions(getExtensions());
    });
  }, [subscribe, getExtensions]);
  const [hostError, setHostError] = React.useState();
  React.useEffect(
    () => host.addEventListener(
      "error",
      (event) => setHostError(event.detail.error)
    ),
    baseDeps
  );
  return { extensions, loading: host.loading, error: hostError };
}
function useExtensionListFetched() {
  const extensionsInfo = React.useContext(ExtensionContext);
  return extensionsInfo.extensionListFetched;
}

exports.Extensible = Extensible;
exports.ExtensionContext = ExtensionContext;
exports.GuestUIFrame = GuestUIFrame;
exports.OutsideOfExtensionContextError = OutsideOfExtensionContextError;
exports.useExtensionListFetched = useExtensionListFetched;
exports.useExtensions = useExtensions;
exports.useHost = useHost;
Object.keys(uixHost).forEach(function (k) {
  if (k !== 'default' && !exports.hasOwnProperty(k)) Object.defineProperty(exports, k, {
    enumerable: true,
    get: function () { return uixHost[k]; }
  });
});
