// @adobe/uix-host/dist/index.js
'use strict';

var uixCore = require('@adobe/uix-core');

// src/host.ts

// src/dom-utils/attribute-normalizers.ts
var tokenizeAttrValues = (tokens) => typeof tokens === "string" ? tokens.split(" ") : tokens;
var mergeAttrValues = (...tokenLists) => {
  const allMerged = /* @__PURE__ */ new Set();
  for (const tokenList of tokenLists) {
    for (const token of tokenizeAttrValues(tokenList)) {
      allMerged.add(token);
    }
  }
  return [...allMerged];
};

// src/dom-utils/iframe-normalizers.ts
var makeSandboxAttrs = (...sandboxes) => mergeAttrValues(...sandboxes);
var requiredIframeProps = {
  // must not require this until app builder supports CSP
  // csp: "frame-ancestors 'self'",
  "data-uix-guest": "true",
  role: "presentation",
  referrerPolicy: "strict-origin"
};
var requiredIframeAttrEntries = Object.entries(requiredIframeProps);
var normalizeIframe = (iframe) => {
  for (const [attr, value] of requiredIframeAttrEntries) {
    iframe.setAttribute(attr, value);
  }
};

// src/port.ts
var defaultOptions = {
  timeout: 1e4,
  debug: false
};
var Port = class extends uixCore.Emitter {
  // #endregion Properties (13)
  // #region Constructors (1)
  constructor(config) {
    super(config.id);
    this.hostApis = {};
    this.isLoaded = false;
    this.subscriptions = [];
    const { timeout, debug } = { ...defaultOptions, ...config.options || {} };
    this.timeout = timeout;
    this.debug = debug;
    this.logger = config.logger;
    this.id = config.id;
    this.url = config.url;
    this.runtimeContainer = config.runtimeContainer;
    this.sharedContext = config.sharedContext;
    this.configuration = config.configuration;
    this.subscriptions.push(
      config.events.addEventListener("contextchange", async (event) => {
        this.sharedContext = event.detail.context;
        await this.load();
        await this.guestServer.getRemoteApi().emit("contextchange", { context: this.sharedContext });
      })
    );
  }
  get apis() {
    if (this.isReady() && this.guestServer) {
      const server = this.guestServer.getRemoteApi();
      return server && server.apis;
    }
  }
  // #endregion Constructors (1)
  // #region Public Methods (6)
  /**
   * Connect an iframe element which is displaying another page in the extension
   * with the extension's bootstrap frame, so they can share context and events.
   */
  attachUI(iframe, privateMethods) {
    return this.attachFrame(iframe, {
      onIframeResize: (dimensions) => {
        this.emit("guestresize", {
          dimensions,
          guestPort: this,
          iframe
        });
      },
      ...privateMethods
    });
  }
  /**
   * Returns true if the guest has registered methods matching the provided
   * capability spec. A capability spec is simply an object whose properties are
   * declared in an array of keys, description the names of the functions and
   * methods that the Port will expose.
   */
  hasCapabilities(requiredCapabilities) {
    this.assertReady();
    return this.apis && Object.entries(requiredCapabilities).every(
      ([apiName, methodNames]) => this.hasCapability(apiName, methodNames)
    );
  }
  /**
   * True when al extensions have loaded.
   */
  isReady() {
    return this.isLoaded && !this.error;
  }
  /**
   * Loads the extension. Returns a promise which resolves when the extension
   * has loaded. The Host calls this method after retrieving extensions.
   */
  async load() {
    try {
      if (!this.isLoaded) {
        await this.connect();
      }
    } catch (e) {
      this.guestServer = null;
      this.error = e instanceof Error ? e : new Error(String(e));
      throw e;
    }
  }
  /**
   * The host-side equivalent of {@link @adobe/uix-guest#register}. Pass a set
   * of methods down to the guest as proxies.
   * Merges at the first level, the API level. Overwrites a deeper levels.
   */
  provide(apis) {
    for (const [apiNamespace, methods] of Object.entries(apis)) {
      this.hostApis[apiNamespace] = this.hostApis[apiNamespace] || {};
      Object.assign(this.hostApis[apiNamespace], methods);
    }
    this.emit("hostprovide", { guestPort: this, apis });
  }
  /**
   * Disconnect from the extension.
   */
  async unload() {
    if (this.guestServerFrame && this.guestServerFrame.parentElement) {
      this.guestServerFrame.parentElement.removeChild(this.guestServerFrame);
      this.guestServerFrame = void 0;
    }
    this.emit("unload", { guestPort: this });
  }
  // #endregion Public Methods (6)
  // #region Private Methods (6)
  hasCapability(apiName, methodNames) {
    const api = this.apis[apiName];
    return api && methodNames.every(
      (methodName) => Reflect.has(api, methodName) && typeof api[methodName] === "function"
    );
  }
  assert(condition, errorMessage) {
    if (!condition) {
      throw new Error(
        `Error in guest extension "${this.id}": ${errorMessage()}`
      );
    }
  }
  assertReady() {
    this.assert(this.isReady(), () => "Attempted to interact before loaded");
  }
  attachFrame(iframe, addedMethods = {}) {
    normalizeIframe(iframe);
    this.logger.log("attachFrame", iframe);
    return uixCore.connectIframe(
      iframe,
      {
        logger: this.logger,
        targetOrigin: this.url.origin,
        timeout: this.timeout
      },
      {
        getSharedContext: () => this.sharedContext,
        getConfiguration: () => this.configuration,
        invokeHostMethod: (address) => this.invokeHostMethod(address, addedMethods),
        ...addedMethods
      }
    );
  }
  async connect() {
    const serverFrame = this.runtimeContainer.ownerDocument.createElement("iframe");
    normalizeIframe(serverFrame);
    serverFrame.setAttribute("aria-hidden", "true");
    serverFrame.setAttribute("src", this.url.href);
    this.guestServerFrame = serverFrame;
    this.runtimeContainer.appendChild(serverFrame);
    if (this.logger) {
      this.logger.info(
        `Guest ${this.id} attached iframe of ${this.url.href}`,
        this
      );
    }
    this.guestServer = await this.attachFrame(serverFrame);
    this.isLoaded = true;
    if (this.logger) {
      this.logger.info(
        `Guest ${this.id} established connection, received methods`,
        this.apis,
        this
      );
    }
  }
  getHostMethodCallee({ name, path }, methodSource) {
    const dots = (level) => `host.${path.slice(0, level).join(".")}`;
    const methodCallee = path.reduce((current, prop, level) => {
      this.assert(
        Reflect.has(current, prop),
        () => `${dots(level)} has no property "${prop}"`
      );
      const next = current[prop];
      this.assert(
        typeof next === "object",
        () => `${dots(
          level
        )}.${prop} is not an object; namespaces must be objects with methods`
      );
      return next;
    }, methodSource);
    this.assert(
      typeof methodCallee[name] === "function" && Reflect.has(methodCallee, name),
      () => `"${dots(path.length - 1)}.${name}" is not a function`
    );
    return methodCallee;
  }
  invokeHostMethod(address, privateMethods) {
    const { name, path, args = [] } = address;
    this.assert(name && typeof name === "string", () => "Method name required");
    this.assert(
      path.length > 0,
      () => `Cannot call a method directly on the host; ".${name}()" must be in a namespace.`
    );
    let methodCallee;
    if (privateMethods) {
      try {
        methodCallee = this.getHostMethodCallee(address, privateMethods);
      } catch (e) {
        this.logger.warn(
          `Private method ${uixCore.formatHostMethodAddress(address)} not found!`,
          address
        );
      }
    }
    if (!methodCallee) {
      methodCallee = this.getHostMethodCallee(address, this.hostApis);
    }
    const method = methodCallee[name];
    this.emit("beforecallhostmethod", { guestPort: this, name, path, args });
    return method.apply(methodCallee, [
      { id: this.id, url: this.url },
      ...args
    ]);
  }
  // #endregion Private Methods (6)
};
function debugHost(host) {
  const hostLogger = uixCore.debugEmitter(host, {
    theme: "blue medium",
    type: "Host"
  });
  hostLogger.listen("guestbeforeload", (log, event) => {
    const { detail } = event;
    const guest = detail.guest;
    log.info(event, `Guest ID ${guest.id}`);
    const portLogger = uixCore.debugEmitter(guest, {
      theme: "green medium",
      type: "Port",
      id: `${host.id} \u2794 ${guest.id}`
    });
    portLogger.listen("hostprovide", (log2, event2) => {
      log2.info("received APIs", event2.detail.apis);
    }).listen("beforecallhostmethod", (log2, event2) => {
      log2.info(event2.detail);
    }).listen("guestresize", (log2, event2) => {
      log2.info(event2.detail);
    }).listen("unload", (log2, event2) => {
      log2.info(event2.detail);
      log2.detach();
    });
  }).listen("guestload", (log, e) => {
    log.info(e.detail.guest.id, e.detail.guest);
  }).listen("error", (log, e) => {
    log.error(`Error: ${e.detail.error.message}`, e);
  }).listen(
    "loadallguests",
    (log, { detail: { failed, loaded, host: host2 } }) => {
      if (failed.length > 0) {
        log.error("%d guests failed to load!", failed.length);
      }
      log.info("%d guests loaded", loaded.length, host2);
    }
  ).listen("unload", (log) => {
    log.info("Unloaded guest and container.");
    log.detach();
  });
  return hostLogger;
}

// src/host.ts
var passAllGuests = () => true;
var _Host = class extends uixCore.Emitter {
  constructor(config) {
    super(config.hostName);
    /**
     * `true` if any extension in {@link Host.guests} has created a {@link
     * @adobe/uix-guest#GuestServer}, but the Guest has not yet loaded.
     */
    this.loading = false;
    /**
     * A Map of of the loaded guests.
     */
    this.guests = /* @__PURE__ */ new Map();
    this.cachedCapabilityLists = /* @__PURE__ */ new WeakMap();
    this.logger = uixCore.quietConsole;
    const { guestOptions = {} } = config;
    this.guestOptions = {
      ...guestOptions,
      debug: guestOptions.debug === false ? false : !!config.debug
    };
    this.hostName = config.hostName;
    this.sharedContext = config.sharedContext || {};
    this.runtimeContainer = config.runtimeContainer;
    if (config.debug) {
      this.logger = debugHost(this);
    }
  }
  getLoadedGuests(filterOrCapabilities) {
    if (typeof filterOrCapabilities === "object") {
      return this.getLoadedGuestsWith(filterOrCapabilities);
    }
    const filter = filterOrCapabilities || passAllGuests;
    const result = [];
    for (const guest of this.guests.values()) {
      if (guest.isReady() && filter(guest)) {
        result.push(guest);
      }
    }
    return result;
  }
  shareContext(setterOrContext) {
    if (typeof setterOrContext === "function") {
      this.sharedContext = setterOrContext(this.sharedContext);
    } else {
      this.sharedContext = setterOrContext;
    }
    this.emit("contextchange", {
      host: this,
      context: this.sharedContext
    });
  }
  /**
   * Load extension into host application from provided extension description.
   * Returned promise resolves when all extensions are loaded and registered.
   *
   * @param extensions - List of extension descriptors. Normally, the Host should receive this value from an {@link ExtensionsProvider}.
   * @param options - Custom options to be used as defaults for each {@link Port} object created for each guest.
   * @returns Promise which resolves when all guests have been loaded.
   */
  async load(extensions, options) {
    this.runtimeContainer = this.runtimeContainer || this.createRuntimeContainer(window);
    const failed = [];
    const loaded = [];
    this.loading = true;
    await Promise.all(
      Object.entries(extensions).map(async ([id, extension]) => {
        const port = await this.loadOneGuest(id, extension, options);
        (port.error ? failed : loaded).push(port);
      })
    );
    this.loading = false;
    this.emit("loadallguests", { host: this, failed, loaded });
  }
  /**
   * Unload all extensions and remove their frames/workers. Use this to unmount
   * a UI or when switching to a different extensible UI.
   */
  async unload() {
    this.emit("beforeunload", { host: this });
    await Promise.all([...this.guests.values()].map((guest) => guest.unload()));
    this.guests.clear();
    this.runtimeContainer.parentElement.removeChild(this.runtimeContainer);
    this.emit("unload", { host: this });
  }
  createRuntimeContainer(window2) {
    const { document } = window2;
    const container = document.createElement("div");
    container.setAttribute("data-uix-guest-container", this.hostName);
    container.setAttribute("role", "presentation");
    container.setAttribute("aria-hidden", "true");
    Object.assign(container.style, _Host.containerStyle);
    document.body.appendChild(container);
    return container;
  }
  async loadOneGuest(id, extension, options = {}) {
    let guest = this.guests.get(id);
    if (!guest) {
      const isExtension = (item) => {
        return typeof item === "object" && item !== null && "url" in item;
      };
      const extensionUrl = isExtension(extension) ? extension.url : extension;
      const extensionConfiguration = isExtension(extension) ? extension.configuration : void 0;
      const url = new URL(extensionUrl);
      guest = new Port({
        owner: this.hostName,
        id,
        url,
        runtimeContainer: this.runtimeContainer,
        options: {
          ...this.guestOptions,
          ...options
        },
        logger: this.logger,
        sharedContext: this.sharedContext,
        configuration: extensionConfiguration,
        events: this
      });
      this.guests.set(id, guest);
    }
    this.emit("guestbeforeload", { guest, host: this });
    try {
      await guest.load();
    } catch (e) {
      const error = new Error(
        `Guest ${guest.id} failed to load: at ${guest.url}: ${e instanceof Error ? e.stack : String(e)}`
      );
      this.emit("error", { host: this, guest, error });
      return guest;
    }
    this.cachedCapabilityLists = /* @__PURE__ */ new WeakMap();
    this.emit("guestload", { guest, host: this });
    return guest;
  }
  getLoadedGuestsWith(capabilities) {
    if (this.cachedCapabilityLists.has(capabilities)) {
      return this.cachedCapabilityLists.get(capabilities);
    }
    const guestsWithCapabilities = this.getLoadedGuests(
      (guest) => guest.hasCapabilities(capabilities)
    );
    this.cachedCapabilityLists.set(capabilities, guestsWithCapabilities);
    return guestsWithCapabilities;
  }
};
var Host = _Host;
Host.containerStyle = {
  position: "fixed",
  width: "1px",
  height: "1px",
  pointerEvents: "none",
  opacity: 0,
  top: 0,
  left: "-1px"
};

// src/extensions-provider/extension-registry.ts
function buildEndpointPath(config) {
  return `${config.service}/${config.extensionPoint}/${config.version}`;
}
function ensureProtocolSpecified(url) {
  if (url.startsWith("https://")) {
    return url;
  }
  if (url.startsWith("http://")) {
    return url;
  }
  return `https://${url}`;
}
async function fetchExtensionsFromRegistry(config) {
  const resp = await fetch(
    `${ensureProtocolSpecified(
      config.baseUrl || "appregistry.adobe.io"
    )}/myxchng/v1/org/${encodeURIComponent(
      config.imsOrg
    )}/xtn/${buildEndpointPath(config)}?auth=true`,
    {
      headers: {
        Accept: "application/json",
        Authorization: `${config.auth.schema} ${config.auth.imsToken}`,
        // todo: check if auth schema needed (initial implementation was without it)
        "X-Api-Key": config.apiKey
      }
    }
  );
  if (resp.status != 200) {
    throw new Error(
      `extension registry returned non-200 response (${resp.status}): ${await resp.text()}`
    );
  }
  return await resp.json();
}
function extensionRegistryExtensionsProvider(config) {
  const erEndpoint = buildEndpointPath(config);
  return fetchExtensionsFromRegistry(config).then(
    (out) => out.reduce((a, e) => {
      if (e.status !== "PUBLISHED") {
        return a;
      }
      return {
        ...a,
        // todo: make safer way to extract href
        [e.name]: e.endpoints[erEndpoint].view[0].href
      };
    }, {})
  );
}
function createExtensionRegistryProvider(config) {
  return function() {
    return extensionRegistryExtensionsProvider(config);
  };
}

// src/extensions-provider/composition.ts
function combineExtensionsFromProviders(...providers) {
  return () => Promise.all(providers.map((ep) => ep())).then(
    (extensionsBatches) => {
      return Object.assign({}, ...extensionsBatches);
    }
  );
}

// src/extensions-provider/mute.ts
function mutedProvider(provider) {
  return async () => {
    try {
      return await provider();
    } catch (error) {
      console.error(`Extension provider has failed: ${error.message}`, {
        error
      });
      return {};
    }
  };
}

exports.Host = Host;
exports.Port = Port;
exports.combineExtensionsFromProviders = combineExtensionsFromProviders;
exports.createExtensionRegistryProvider = createExtensionRegistryProvider;
exports.makeSandboxAttrs = makeSandboxAttrs;
exports.mergeAttrValues = mergeAttrValues;
exports.mutedProvider = mutedProvider;
exports.normalizeIframe = normalizeIframe;
exports.requiredIframeProps = requiredIframeProps;
exports.tokenizeAttrValues = tokenizeAttrValues;
